@weeco thanks.

First of all you have some linting errors:
```
# github.com/grafana/grafana/pkg/models [github.com/grafana/grafana/pkg/models.test]
pkg/models/datasource_cache_test.go:34:9: t1.TLSClientConfig undefined (type http.RoundTripper has no field or method TLSClientConfig)
pkg/models/datasource_cache_test.go:37:13: t1.TLSClientConfig undefined (type http.RoundTripper has no field or method TLSClientConfig)
pkg/models/datasource_cache_test.go:40:9: t1.TLSClientConfig undefined (type http.RoundTripper has no field or method TLSClientConfig)
pkg/models/datasource_cache_test.go:65:9: t1.TLSClientConfig undefined (type http.RoundTripper has no field or method TLSClientConfig)
pkg/models/datasource_cache_test.go:68:13: t1.TLSClientConfig undefined (type http.RoundTripper has no field or method TLSClientConfig)
pkg/models/datasource_cache_test.go:71:9: t1.TLSClientConfig undefined (type http.RoundTripper has no field or method TLSClientConfig)
pkg/models/datasource_cache_test.go:82:9: t2.TLSClientConfig undefined (type http.RoundTripper has no field or method TLSClientConfig)
pkg/models/datasource_cache_test.go:113:9: tr.TLSClientConfig undefined (type http.RoundTripper has no field or method TLSClientConfig)
pkg/models/datasource_cache_test.go:116:13: tr.TLSClientConfig undefined (type http.RoundTripper has no field or method TLSClientConfig)
pkg/models/datasource_cache_test.go:142:9: tr.TLSClientConfig undefined (type http.RoundTripper has no field or method TLSClientConfig)
pkg/models/datasource_cache_test.go:142:9: too many errors
make: *** [Makefile:111: go-vet] Error 2
Exited with code 2
```

Then you have some nil-pointer errors: https://circleci.com/gh/grafana/grafana/159848?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-checks-link

--- a/pkg/api/pluginproxy/ds_proxy_test.go
+++ b/pkg/api/pluginproxy/ds_proxy_test.go
@@ -17,7 +17,6 @@ import (
 
 	"github.com/grafana/grafana/pkg/bus"
 	"github.com/grafana/grafana/pkg/components/simplejson"
-	"github.com/grafana/grafana/pkg/infra/log"
 	"github.com/grafana/grafana/pkg/login/social"
 	m "github.com/grafana/grafana/pkg/models"
 	"github.com/grafana/grafana/pkg/plugins"
@@ -328,37 +327,6 @@ func TestDSRouteRule(t *testing.T) {
 
 			Convey("Should keep named cookies", func() {
 				So(req.Header.Get("Cookie"), ShouldEqual, "JSESSION_ID=test")
-			})
-		})
-
-		Convey("When proxying a data source with custom headers specified", func() {
-			plugin := &plugins.DataSourcePlugin{}
-
-			encryptedData, err := util.Encrypt([]byte(`Bearer xf5yhfkpsnmgo`), setting.SecretKey)
-			ds := &m.DataSource{
-				Type: m.DS_PROMETHEUS,
-				Url:  "http://prometheus:9090",
-				JsonData: simplejson.NewFromAny(map[string]interface{}{
-					"httpHeaderName1": "Authorization",
-				}),
-				SecureJsonData: map[string][]byte{
-					"httpHeaderValue1": encryptedData,
-				},
-			}
-
-			ctx := &m.ReqContext{}
-			proxy := NewDataSourceProxy(ds, plugin, ctx, "", &setting.Cfg{})
-
-			requestURL, _ := url.Parse("http://grafana.com/sub")
-			req := http.Request{URL: requestURL, Header: make(http.Header)}
-			proxy.getDirector()(&req)
-
-			if err != nil {
-				log.Fatal(4, err.Error())
-			}
-
-			Convey("Match header value after decryption", func() {
-				So(req.Header.Get("Authorization"), ShouldEqual, "Bearer xf5yhfkpsnmgo")
 			})
 		})
 

--- a/pkg/models/datasource_cache.go
+++ b/pkg/models/datasource_cache.go
@@ -18,23 +18,25 @@ type proxyTransportCache struct {
 	sync.Mutex
 }
 
-type headerRoundTripper struct {
-	headers map[string]string
-	proxied http.RoundTripper
+// dataSourceTransport implements http.RoundTripper (https://golang.org/pkg/net/http/#RoundTripper)
+type dataSourceTransport struct {
+	headers   map[string]string
+	transport *http.Transport
 }
 
-func (h headerRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
-	for key, value := range h.headers {
+// RoundTrip executes a single HTTP transaction, returning a Response for the provided Request.
+func (d *dataSourceTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	for key, value := range d.headers {
 		req.Header.Set(key, value)
 	}
 
-	return h.proxied.RoundTrip(req)
+	return d.transport.RoundTrip(req)
 }
 
 type cachedTransport struct {
 	updated time.Time
 
-	Transport http.RoundTripper
+	*dataSourceTransport
 }
 
 var ptc = proxyTransportCache{
@@ -54,12 +56,12 @@ func (ds *DataSource) GetHttpClient() (*http.Client, error) {
 	}, nil
 }
 
-func (ds *DataSource) GetHttpTransport() (http.RoundTripper, error) {
+func (ds *DataSource) GetHttpTransport() (*dataSourceTransport, error) {
 	ptc.Lock()
 	defer ptc.Unlock()
 
 	if t, present := ptc.cache[ds.Id]; present && ds.Updated.Equal(t.updated) {
-		return t.Transport, nil
+		return t.dataSourceTransport, nil
 	}
 
 	tlsConfig, err := ds.GetTLSConfig()
@@ -83,17 +85,18 @@ func (ds *DataSource) GetHttpTransport() (http.RoundTripper, error) {
 		MaxIdleConns:          100,
 		IdleConnTimeout:       90 * time.Second,
 	}
-	proxiedTransport := &headerRoundTripper{
-		proxied: transport,
-		headers: customHeaders,
+
+	dsTransport := &dataSourceTransport{
+		headers:   customHeaders,
+		transport: transport,
 	}
 
 	ptc.cache[ds.Id] = cachedTransport{
-		Transport: proxiedTransport,
-		updated:   ds.Updated,
+		dataSourceTransport: dsTransport,
+		updated:             ds.Updated,
 	}
 
-	return proxiedTransport, nil
+	return dsTransport, nil
 }
 
 func (ds *DataSource) GetTLSConfig() (*tls.Config, error) {
@@ -135,6 +138,9 @@ func (ds *DataSource) GetTLSConfig() (*tls.Config, error) {
 // The map key represents the HeaderName and the value represents this header's value
 func (ds *DataSource) getCustomHeaders() map[string]string {
 	headers := make(map[string]string)
+	if ds.JsonData == nil {
+		return headers
+	}
 
 	decrypted := ds.SecureJsonData.Decrypt()
 	index := 1

--- a/pkg/models/datasource_cache_test.go
+++ b/pkg/models/datasource_cache_test.go
@@ -1,6 +1,7 @@
 package models
 
 import (
+	"log"
 	"testing"
 	"time"
 
@@ -164,6 +165,33 @@ func TestDataSourceCache(t *testing.T) {
 
 		Convey("Should skip TLS verification", func() {
 			So(tr.transport.TLSClientConfig.InsecureSkipVerify, ShouldEqual, true)
+		})
+	})
+
+	Convey("When caching a datasource proxy with custom headers specified", t, func() {
+		clearCache()
+
+		encryptedData, err := util.Encrypt([]byte(`Bearer xf5yhfkpsnmgo`), setting.SecretKey)
+		if err != nil {
+			log.Fatal(err.Error())
+		}
+
+		json := simplejson.NewFromAny(map[string]interface{}{
+			"httpHeaderName1": "Authorization",
+		})
+
+		ds := DataSource{
+			Id:             1,
+			Url:            "http://k8s:8001",
+			Type:           "Kubernetes",
+			JsonData:       json,
+			SecureJsonData: map[string][]byte{"httpHeaderValue1": encryptedData},
+		}
+
+		headers := ds.getCustomHeaders()
+
+		Convey("Should match header value after decryption", func() {
+			So(headers["httpHeaderName1"], ShouldEqual, "Bearer xf5yhfkpsnmgo")
 		})
 	})
 }

--- a/pkg/models/datasource_cache_test.go
+++ b/pkg/models/datasource_cache_test.go
@@ -191,7 +191,7 @@ func TestDataSourceCache(t *testing.T) {
 		headers := ds.getCustomHeaders()
 
 		Convey("Should match header value after decryption", func() {
-			So(headers["httpHeaderName1"], ShouldEqual, "Bearer xf5yhfkpsnmgo")
+			So(headers["Authorization"], ShouldEqual, "Bearer xf5yhfkpsnmgo")
 		})
 	})
 }

--- a/pkg/models/datasource_cache_test.go
+++ b/pkg/models/datasource_cache_test.go
@@ -31,13 +31,13 @@ func TestDataSourceCache(t *testing.T) {
 			So(t2, ShouldEqual, t1)
 		})
 		Convey("Should verify TLS by default", func() {
-			So(t1.TLSClientConfig.InsecureSkipVerify, ShouldEqual, false)
+			So(t1.transport.TLSClientConfig.InsecureSkipVerify, ShouldEqual, false)
 		})
 		Convey("Should have no TLS client certificate configured", func() {
-			So(len(t1.TLSClientConfig.Certificates), ShouldEqual, 0)
+			So(len(t1.transport.TLSClientConfig.Certificates), ShouldEqual, 0)
 		})
 		Convey("Should have no user-supplied TLS CA onfigured", func() {
-			So(t1.TLSClientConfig.RootCAs, ShouldBeNil)
+			So(t1.transport.TLSClientConfig.RootCAs, ShouldBeNil)
 		})
 	})
 
@@ -62,13 +62,13 @@ func TestDataSourceCache(t *testing.T) {
 		So(err, ShouldBeNil)
 
 		Convey("Should verify TLS by default", func() {
-			So(t1.TLSClientConfig.InsecureSkipVerify, ShouldEqual, false)
+			So(t1.transport.TLSClientConfig.InsecureSkipVerify, ShouldEqual, false)
 		})
 		Convey("Should have no TLS client certificate configured", func() {
-			So(len(t1.TLSClientConfig.Certificates), ShouldEqual, 0)
+			So(len(t1.transport.TLSClientConfig.Certificates), ShouldEqual, 0)
 		})
 		Convey("Should have no user-supplied TLS CA configured", func() {
-			So(t1.TLSClientConfig.RootCAs, ShouldBeNil)
+			So(t1.transport.TLSClientConfig.RootCAs, ShouldBeNil)
 		})
 
 		ds.JsonData = nil
@@ -79,7 +79,7 @@ func TestDataSourceCache(t *testing.T) {
 		So(err, ShouldBeNil)
 
 		Convey("Should have no user-supplied TLS CA configured after the update", func() {
-			So(t2.TLSClientConfig.RootCAs, ShouldBeNil)
+			So(t2.transport.TLSClientConfig.RootCAs, ShouldBeNil)
 		})
 	})
 
@@ -110,10 +110,10 @@ func TestDataSourceCache(t *testing.T) {
 		So(err, ShouldBeNil)
 
 		Convey("Should verify TLS by default", func() {
-			So(tr.TLSClientConfig.InsecureSkipVerify, ShouldEqual, false)
+			So(tr.transport.TLSClientConfig.InsecureSkipVerify, ShouldEqual, false)
 		})
 		Convey("Should have a TLS client certificate configured", func() {
-			So(len(tr.TLSClientConfig.Certificates), ShouldEqual, 1)
+			So(len(tr.transport.TLSClientConfig.Certificates), ShouldEqual, 1)
 		})
 	})
 
@@ -139,10 +139,10 @@ func TestDataSourceCache(t *testing.T) {
 		So(err, ShouldBeNil)
 
 		Convey("Should verify TLS by default", func() {
-			So(tr.TLSClientConfig.InsecureSkipVerify, ShouldEqual, false)
+			So(tr.transport.TLSClientConfig.InsecureSkipVerify, ShouldEqual, false)
 		})
 		Convey("Should have a TLS CA configured", func() {
-			So(len(tr.TLSClientConfig.RootCAs.Subjects()), ShouldEqual, 1)
+			So(len(tr.transport.TLSClientConfig.RootCAs.Subjects()), ShouldEqual, 1)
 		})
 	})
 
@@ -163,7 +163,7 @@ func TestDataSourceCache(t *testing.T) {
 		So(err, ShouldBeNil)
 
 		Convey("Should skip TLS verification", func() {
-			So(tr.TLSClientConfig.InsecureSkipVerify, ShouldEqual, true)
+			So(tr.transport.TLSClientConfig.InsecureSkipVerify, ShouldEqual, true)
 		})
 	})
 }

